Fixes #15527 - api/v2/hosts is slow loading permissions

dLobatog opened the pull request:
The .eager_load call when called with [] generates a SELECT query that
selects all host attributes. This can cause it to take a lot longer when
there are large number of hosts involved. These attributes are
irrelevant in this context (when they are relevant, they're put in the
scope_components hash)

With 5700 hosts, it takes about 50% less time to run this than to run it
with eager_load.

This is the query I mean:

``` sql
SELECT `hosts`.`id` AS t0_r0, `hosts`.`name` AS t0_r1, `hosts`.`last_compile` AS t0_r2, `hosts`.`last_report` AS t0_r3, `hosts`.`updated_at` AS t0_r4, `hosts`.`created_at` AS t0_r5, `hosts`.`root_pass` AS t0_r6, `hosts`.`architecture_id` AS t0_r7, `hosts`.`operatingsystem_id` AS t0_r8, `hosts`.`environment_id` AS t0_r9, `hosts`.`ptable_id` AS t0_r10, `hosts`.`medium_id` AS t0_r11, `hosts`.`build` AS t0_r12, `hosts`.`comment` AS t0_r13, `hosts`.`disk` AS t0_r14, `hosts`.`installed_at` AS t0_r15, `hosts`.`model_id` AS t0_r16, `hosts`.`hostgroup_id` AS t0_r17, `hosts`.`owner_id` AS t0_r18, `hosts`.`owner_type` AS t0_r19, `hosts`.`enabled` AS t0_r20, `hosts`.`puppet_ca_proxy_id` AS t0_r21, `hosts`.`managed` AS t0_r22, `hosts`.`use_image` AS t0_r23, `hosts`.`image_file` AS t0_r24, `hosts`.`uuid` AS t0_r25, `hosts`.`compute_resource_id` AS t0_r26, `hosts`.`puppet_proxy_id` AS t0_r27, `hosts`.`certname` AS t0_r28, `hosts`.`image_id` AS t0_r29, `hosts`.`organization_id` AS t0_r30, `hosts`.`location_id` AS t0_r31, `hosts`.`type` AS t0_r32, `hosts`.`otp` AS t0_r33, `hosts`.`realm_id` AS t0_r34, `hosts`.`compute_profile_id` AS t0_r35, `hosts`.`provision_method` AS t0_r36, `hosts`.`grub_pass` AS t0_r37, `hosts`.`salt_proxy_id` AS t0_r38, `hosts`.`salt_environment_id` AS t0_r39, `hosts`.`discovery_rule_id` AS t0_r40, `hosts`.`global_status` AS t0_r41, `hosts`.`lookup_value_matcher` AS t0_r42, `hosts`.`chef_proxy_id` AS t0_r43 FROM `hosts` WHERE `hosts`.`type` IN ('Host::Managed')
```

vs just 

``` sql
SELECT `hosts`.* FROM `hosts` WHERE `hosts`.`type` IN ('Host::Managed')
```

In order to reproduce the issue, run the following script with a non-admin user with several roles, who's able to see many hosts (after 5k or so it becomes obviously slow)

```
  User.current = User.find(134)
  authorizer = Authorizer.new(User.current)
  host = Host.last
  puts Benchmark.measure {
    Permission.where(:resource_type => "Host").all.inject({}) do |hash, permission|                                                           puts permission
      puts Benchmark.measure { authorizer.can?(permission.name, host) }
    end
    puts 'total time'
  }
```

I've created this file to benchmark a few more things along the way, but it's tough since Rails delays running the query and chains them, anyone knows how to make it run the queries right away?

dLobatog committed:
Fixes #15527 - api/v2/hosts is slow loading permissions

The .eager_load call when called with [] generates a SELECT query that
selects all host attributes. This can cause it to take a lot longer
when there are large number of hosts involved. These attributes are
irrelevant in this context (when they are relevant, they're put in the
scope_components hash)

With 5700 hosts, it takes about 50% less time to run this than to run
it with eager_load.

Chaining the call with .where makes the call near-instant.
It selects the host from the list of hosts we can see using the db
directly.

domcleal commented on app/services/authorizer.rb.backup:
```diff
@@ -0,0 +1,157 @@
+class Authorizer
```
This backup file shouldn't be present.

domcleal commented:
> SELECT `hosts`.* FROM `hosts` WHERE `hosts`.`type` IN ('Host::Managed')

There are many repetitions of this query for every different permission, which is probably why this is so inefficient. It may help if the

domcleal commented:
> SELECT `hosts`.\* FROM `hosts` WHERE `hosts`.`type` IN ('Host::Managed')

There are many repetitions of this query for every different permission, which is probably why this is so inefficient. It's optimised for checking one permission across many hosts rather than checking many permissions for one host, which is why the per-permission collection is cached so #include? can be called for each host.

It may help to add a new option to the `can?` method that optimises it for the case where the Authorizer will only be used to check for one resource - i.e. you call `find_collection(..).where(:id => subject.id)` or similar to change the retrieval of _all_ hosts to a single one.

Sorry for the duplicate comment, I clicked Comment by accident.

dLobatog commented:
@domcleal That indeed makes a huge difference, still the eager_load removes unnecessary params from the query (we don't need to know anything other than the ID to complete it). Thanks!

dLobatog commented on app/services/authorizer.rb:
```diff
@@ -12,8 +12,9 @@ def can?(permission, subject = nil)
       return true if user.admin?
-      collection = @cache[subject.class.to_s][permission] ||= find_collection(subject.class, :permission => permission)
-      collection.include?(subject)
+      @cache["#{subject.class.to_s}-#{permission}-#{subject.id}"] ||=
```
I've changed this hash to be just 1 level instead of nested as there was no purpose for anything but the 'leaf' nodes afaik

beav replied:
How does the cache get invalidated, for example when roles are altered?

dLobatog replied:
It depends on the action, for most actions a new Authorizer is created so the cache is thrown away (https://github.com/theforeman/foreman/blob/develop/app/helpers/application_helper.rb#L104)

dLobatog replied:
@ares if you know this well - let me know here and I'll update it with a comment (or even a new PR would help)

domcleal commented on app/services/authorizer.rb:
```diff
@@ -12,8 +12,9 @@ def can?(permission, subject = nil)
-      collection.include?(subject)
+      @cache["#{subject.class.to_s}-#{permission}-#{subject.id}"] ||=
+        find_collection(subject.class, :permission => permission).
+        where(:id => subject.id)
```
I think this new `where(:id..)` behaviour should be optional, keeping the original caching of the whole resource list for the regular case of one permission, multiple resources - e.g. permission checking a button on an index page.  A per-resource cache will be ineffective when the resource ID changes, so it'll generate 2*N SQL queries or similar on those pages.

Perhaps add an options hash to the `can?` method to specify the optimal caching strategy?

Edit: probably something to change separately, but perhaps the regular permissions check can be optimised further by only plucking an array of IDs from the DB rather than full models.

dLobatog replied:
Yeah you're right - although these queries are considerably faster now that I wonder whether the cache is even necessary now. I did it this way because storing the collection in the cache would mean having to retrieve collection (without `where`), which was the bottleneck.

I'll see how can I signal to this method that I don't want to get `find_collection`, maybe with an option hash as you said !

domcleal replied:
Yes, I do partly agree that the collection may be obsolete or could be optimised further, though removing it now will require more care to avoid continually reloading filters, permissions etc on index pages.  I'd prefer to look at that separately as it's optimising for the opposite (and more common) case to this.

nbarrientos commented:
Hi,

Is there any chance this patch gets reviewed soon? We're looking forward to backporting it as the performance drop that we got after upgrading to 1.11.2 is biting us (+20k hosts).

Thank you!

domcleal commented:
@dLobatog thanks for the update, please don't forget to leave a comment, I didn't know the PR had changed (ta @nbarrientos).

Looks fine now, but could you just add a bit of unit test coverage to test/unit/authorizer_test.rb for the third cache argument to `#can?`?  Perhaps run the same set of tests for `[true, false].each do |cache|` since they're two different implementations?

dLobatog commented:
Updated now with tests for cache = true / false. @domcleal I'm seeing GitHub sends me a reminder email when a commit ID changes in a PR (since a few weeks ago), that's why I didn't comment, sorry.

The tests unearthed a few cases where the changes I made didn't make sense, because  the last line of `can?` (https://github.com/theforeman/foreman/pull/3618/files#diff-2a8c1ced7ef4dfe08f35154c5a3ef0f1R19) returned a collection. I added `.present?` to make it return a boolean.

domcleal commented on app/services/authorizer.rb:
```diff
@@ -1,19 +1,23 @@
+      return collection_cache_lookup(subject, permission) if cache
+
+      find_collection(subject.class, :permission => permission).
+        where(:id => subject.id).present?
```
Prefer `any?` to `present?` when dealing with presence and AR collections, it optimises it using a COUNT so doesn't need to retrieve the contents of the columns.  This currently does:

```
2016-07-12T09:43:44 [sql] [D]   Host::Managed Load (0.2ms)  SELECT "hosts".* FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("hosts"."name" = 'wstest.example.com')) AND "hosts"."id" = ?  [["id", 118]]
```

but with #any? it does:

```
2016-07-12T09:49:12 [sql] [D]    (0.2ms)  SELECT COUNT(*) FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("hosts"."name" = 'wstest.example.com')) AND "hosts"."id" = ?  [["id", 118]]
```

dLobatog replied:
Nice optimization, thanks

domcleal commented on test/unit/authorizer_test.rb:
```diff
@@ -10,169 +10,201 @@ def setup
+        refute auth.can?(:view_domains, nil, cache)
+      end
+
+      test "#can?(:view_hosts, nil, cache) is limited by particular user - #{cache}" do
```
Nitpick, use a `context "with cache=#{cache}"` or similar to save changing all of these, or put the value of cache in the method signature at the start of this test name, e.g. `#can?(:view_hosts, nil, #{cache})` so it's clearer what the true/false at the end of the line is.

domcleal commented:
> Updated now with tests for cache = true / false. @domcleal I'm seeing GitHub sends me a reminder email when a commit ID changes in a PR (since a few weeks ago), that's why I didn't comment, sorry.

No worries, I think that only happens now when new commits are added and not if the branch is force-pushed.

dLobatog commented on test/unit/authorizer_test.rb:
```diff
@@ -10,169 +10,201 @@ def setup
-    refute auth.can?(:view_domains)
-  end
-
-  test "#can?(:view_hosts) with unlimited filter" do
```
This had the same signature as the previous test, huh..

domcleal replied:
It should probably say "with wildcard filter". The previous test has a nil/unlimited search, this has `name = *`.

The duplicate name shouldn't be an issue in minitest with `test`, only when using `def` (i.e. they'll be uniquely numbered/named, both should still run).

dLobatog commented:
Updated to include the `.any? instead of`.present?`optimization after WHERE. Notice the test signatures now do not include`can?`as that's already in the 'describe' block. I made a`context` to  show whether cache was true or false

domcleal commented on test/unit/authorizer_test.rb:
```diff
@@ -10,169 +10,203 @@ def setup
+          refute auth.can?(@permission.name.to_sym, nil, cache)
+        end
+
+        test "for unlimited filter" do
```
These tests that passed a second subject argument through have lost that info, they previously showed `#can?(:view_domains, @host)` which is different to the "with unlimited filter" test above.  It's not very clear from the test name now, even with the describe/context blocks.

dLobatog replied:
Fair enough, I'm modifying these to say `subject (Domain) provided`

dLobatog replied:
Or I'll put them in a `subject provided` context actually.

dLobatog commented:
Updated tests, they're on a different context depending on whether a subject is provided to .can or not

domcleal commented:
[test] as the [job](http://ci.theforeman.org/job/test_develop_pull_request/15856/) timed out.

domcleal commented on test/unit/authorizer_test.rb:
```diff
@@ -10,169 +10,207 @@ def setup
+              with(:id => domain2.id).returns([])
+            domains_collection.stubs(:where).
+              with(:id => architecture.id).returns([])
+            auth.stubs(:find_collection).returns(domains_collection).at_most(6)
```
Using `at_most` here is going to stop this testing that the cache works when `cache=true` because it'll make 6 calls like `cache=false`.  Perhaps go back to `times` by doing: `times(cache ? 3 : 6)`?

domcleal commented on test/unit/authorizer_test.rb:
```diff
@@ -10,169 +10,207 @@ def setup
+          end
+        end
+
+        context 'with subject (e.g: Domain)' do
```
nitpick, s/Domain/Host/

dLobatog replied:
The subject we're passing to `can?` is `Domain` in the tests below

ares commented on app/services/authorizer.rb:
```diff
@@ -1,19 +1,23 @@
 class Authorizer
+  include AuthorizerCache
```
Why is this extra module? I don't see it being included anywhere else. If you want to extract it from `Authorizer` to keep the object responsibility simple, I'd recommend extracting it to a separate class under Authorizer name scope.

ares commented on app/services/authorizer.rb:
```diff
@@ -1,19 +1,23 @@
       return true if user.admin?
-      collection = @cache[subject.class.to_s][permission] ||= find_collection(subject.class, :permission => permission)
-      collection.include?(subject)
+      return collection_cache_lookup(subject, permission) if cache
```
Since `user` is public accessor it can be changed for existing authorizer object, cache does not take user into account. I think to prevent security issues in future it should either invalidate the cache after user change or disable the changing of user.

dLobatog replied:
I'll do the latter and make user just an `attr_reader`

dLobatog commented:
@ares Updated so that the user cannot be changed for an authorizer, and AuthorizerCache is a class instead of a module

nbarrientos commented:
We've backported the patch (the way it was at the time of writing these lines) into a development instance that has a snapshot of the production data.

When doing:

`GET api/hosts/foo.example.org`:

Before:

```
real    0m11.508s
user    0m0.128s
sys     0m0.008s
```

After:

```
real    0m0.711s
user    0m0.128s
sys     0m0.012s
```

Thanks.

dLobatog commented:
@ares I reverted the  class `AuthorizedCache` and made it a module again. The reason is because I'm calling `find_collection` in there, which I cannot access from the cache class. The change I made to make `:user` a read-only attribute is still there.

domcleal commented:
Merged as 1ff7e25, thanks @dLobatog.
